File: cpp_src/puppy_core/fall_event_adapter.cpp
// Fall Event Adapter 实现
#include "puppy_core/fall_event_adapter.h"

namespace puppy_core {

FallEventAdapterNode::FallEventAdapterNode()
    : rclcpp::Node("fall_event_adapter") {
  pub_ = this->create_publisher<puppy_interfaces::msg::FallEvent>(
      "/fall/event", 10);
  sub_ = this->create_subscription<std_msgs::msg::Bool>(
      "/fall_detected", 10,
      std::bind(&FallEventAdapterNode::onFallDetected, this,
                std::placeholders::_1));

  RCLCPP_INFO(this->get_logger(), "Fall event adapter started");
}

void FallEventAdapterNode::onFallDetected(
    const std_msgs::msg::Bool::SharedPtr msg) {
  puppy_interfaces::msg::FallEvent event;
  event.header.stamp = this->now();
  event.detected = msg->data;
  event.direction = "unknown";
  event.confidence = msg->data ? 1.0f : 0.0f;
  event.source = "legacy_bool_adapter";
  pub_->publish(event);
}

}  // namespace puppy_core

File: src/puppy_core/puppy_core/fall_event_adapter.py
"""Adapt legacy /fall_detected Bool topic into semantic FallEvent."""
import rclpy
from rclpy.node import Node
from std_msgs.msg import Bool
from puppy_interfaces.msg import FallEvent


class FallEventAdapterNode(Node):
    """Convert legacy Bool fall signal into puppy_interfaces/FallEvent."""

    def __init__(self):
        super().__init__('fall_event_adapter')

        self.pub = self.create_publisher(FallEvent, '/fall/event', 10)
        self.create_subscription(Bool, '/fall_detected', self._on_fall_detected, 10)

        self.get_logger().info('Fall event adapter started')

    def _on_fall_detected(self, msg: Bool):
        event = FallEvent()
        event.header.stamp = self.get_clock().now().to_msg()
        event.detected = bool(msg.data)
        event.direction = 'unknown'
        event.confidence = 1.0 if msg.data else 0.0
        event.source = 'legacy_bool_adapter'
        self.pub.publish(event)


def main(args=None):
    rclpy.init(args=args)
    node = FallEventAdapterNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()
